### PR TITLE
Get rid of ATL dependency

### PR DESCRIPTION
**Pull request by d0vgan**
The ATL dependency is quite annoying when working in Visual Studio Express (where there's no ATL included). Moreover, currently it's easy to get rid of the ATL dependency at all - actually it does not add anything to the code that really worth this dependency.

### CODE REVIEW

File: PowerEditor/src/WinControls/ReadDirectoryChanges/ThreadSafeQueue.h [hunk 1]
```diff
@@ -28,6 +28,70 @@
 
 #include <list>
 
+class CCriticalSection
```
Comment (d0vgan) [inline PowerEditor/src/WinControls/ReadDirectoryChanges/ThreadSafeQueue.h]: A very intuitive class, with its implementation quite similar to ATL's CComAutoCriticalSection. (Actually, there is nothing "Com" in it.)

File: PowerEditor/src/WinControls/ReadDirectoryChanges/ThreadSafeQueue.h [hunk 2]
```diff
@@ -28,6 +28,70 @@
...
+        ::LeaveCriticalSection(&m_cs);
+    }
+
+    CCriticalSection(const CCriticalSection&) = delete;
+    CCriticalSection* operator=(const CCriticalSection&) = delete;
+protected:
+    CRITICAL_SECTION m_cs;
+};
+
+class CCriticalSectionLock
```
Comment (d0vgan) [inline PowerEditor/src/WinControls/ReadDirectoryChanges/ThreadSafeQueue.h]: The implementation is quite similar to ATL's CComCritSecLock.

**Comment by MAPJe71** [comment]
Why implement a new locking mechanism when there is already one available in `.\src\MISC\common\mutex.*`?

**Comment by d0vgan** [comment]
Agree, Yuni::Mutex could be used instead. In such case I'll add a new class Yuni::MutexLockerEx that will include the m_bLocked member and additional methods Lock and Unlock - similar to CComCritSecLock.

**Comment by MAPJe71** [comment]
There's already a Yuni::MutexLocker class. Why not use that one?

**Comment by d0vgan** [comment]
@MAPJe71, please look at the implementation of Yuni::MutexLocker and then re-read my previous comment :) This is because of the following line in CThreadSafeQueue::push :
lock.Unlock();
So it's kind of lock_guard with additional manual methods Lock and Unlock.

**Comment by MAPJe71** [comment]
Dispite using an autolocker/lock-guard on a mutex you can still manually (un-)lock it e.g.
```C++
template <typename C>
class CThreadSafeQueue : protected std::list<C>
{
public:

	void push(C& c)
	{
		Yuni::MutexLocker lock(m_Mutex); //<-- auto lock (RAII)
		push_back( c );
		m_Mutex.unlock(); //<-- manual unlock

		if (!::ReleaseSemaphore(m_hSemaphore, 1, NULL))
		{
			// If the semaphore is full, then take back the entry.
			m_Mutex.lock(); //<-- manual lock
			pop_back();
			if (GetLastError() == ERROR_TOO_MANY_POSTS)
			{
				m_bOverflow = true;
			}
		}
	} //<-- auto unlock

	bool pop(C& c)
	{
		Yuni::MutexLocker lock(m_Mutex);

        // 
	}

protected:
	Yuni::Mutex m_Mutex;
};
```

**Comment by d0vgan** [comment]
@MAPJe71,
I bet you've already noticed what's wrong with that code.
Surely, it's this situation:
~~~
{
    Yuni::MutexLocker lock(m_Mutex); //<-- auto lock (RAII)
    m_Mutex.unlock(); //<-- manual unlock
} //<-- auto unlock - shit, unlocked twice!
~~~

**Comment by MAPJe71** [comment]
Oooh, d#mn it you're right that won't work (as intended) 😁

@donho and @milipili could you extend the Yuni::MutexLocker class with manual un-/lock?
Would you consider a PR for [libyuni](https://github.com/libyuni/libyuni)?

**Comment by d0vgan** [comment]
I'd rather propose to have 2 classes: MutexLocker and MutexLockerEx, without inheritance.
Here is why:
1) MutexLockerEx will have an additional bool variable (see the proposed CCriticalSectionLock implementation), which is simply not needed for a lightweight MutexLocker.
2) MutexLockerEx will have an additional argument in the constructor (see the proposed CCriticalSectionLock implementation), which is simply not needed for a lightweight MutexLocker.

**Comment by d0vgan** [comment]
OK, guys - no additional classes now. Just a little thinking and a little bit more braces :)

**Comment by d0vgan** [comment]
Now I believe it's good enough.

**Comment by chcg** [comment]
@d0vgan additional benefit: building with mingw would be possible again, because ATL is not available there, see  #4145 and there #2478 with patches to avoid ATL, see https://github.com/chcg/notepad-plus-plus/commit/463bdc5e1416a5f2ee256d7316ed49e4865a0171#diff-5b536d07367d39b369aea9582ff0fe9d

with std::unique_lock<std::mutex> as replacement for CComCritSecLock<CComAutoCriticalSection>

**Comment by d0vgan** [comment]
rebased + now everything in 1 commit

**Comment by d0vgan** [comment]
Come on, why these ATL things are still here?
"Entities are not to be multiplied without necessity"!

**Comment by d0vgan** [comment]
I used to think that it's me who is conservative and not fast - now I see I'm not alone :)

**Comment by milipili** [comment]
@MAPJe71 From my point of view and from my experience, unlocking a mutex in the middle of the code where it is supposed to be managed by the scope is far from a good idea. Like deleting some resource managed by `unique_ptr`. For this reason, it is unlikely that an `unlock` mecanism comes in libyuni, better fix the code with proper raii (but feel free to bring some particular use-case).

Regarding libyuni/mutex, this is historical reason as `std::mutex` didn't exist before as you can guess. As of now, as the official minimum version for n++ is 2015 (if not mistaken, would be better to switch to 2017 from my point of view, there is a better support for multithread and brings multiple fixes), I would recommend to replace yuni/mutex by its stl counterpart as there is no added value.

That's said, @d0vgan is right using yuni/mutex here, for code consistency. That replacement has nothing to do with this MR and should be another one.

**Comment by milipili** [comment]
@d0vgan I would encourage you next time to use multiple commits as it changes different things.

**Comment by milipili** [comment]
@MAPJe71 Where did you find the snippet you posted regarding `CThreadSafeQueue`? I am surely missing something here as I can't find the `unlock()`

**Comment by milipili** [comment]
@donho For information, libyuni is going full C++17, so `Yuni::Mutex`should disappear in the future (depending if users are still using it or not)

**Comment by milipili** [comment]
@donho This class `CThreadSafeQueue` should die, or at least rewritten. No one must inherit from any stl container, as specified by the standard. Better use encapsulation instead of inheritance anyway.

**Comment by milipili** [comment]
ATL dependency is really annoying. Nice job :)

**Comment by d0vgan** [comment]
After some thinking, I think I vote for chcg's patch:
chcg/notepad-plus-plus@463bdc5#diff-5b536d07367d39b369aea9582ff0fe9d
It uses the standard STL's std::mutex - and it is better than relying on thirdparty's Yuni::MutexLocker, which additionally needs to be consireder to be updated.

**Comment by MIvanchev** [comment]
@d0vgan: just to provide some aux. info, since I had the same fundamental problem as well. ATL/MFC is available with the [Visual C++ Build Tools 2015](https://blogs.msdn.microsoft.com/vcblog/2016/03/31/announcing-the-official-release-of-the-visual-c-build-tools-2015/). You can build Notepad++ easily that way.

**Comment by chcg** [comment]
@MIvanchev The main point to get ride of ATL I think is to be able to build this under mingw.

**Comment by MIvanchev** [comment]
> @MIvanchev The main point to get ride of ATL I think is to be able to build this under mingw.

Yes, I also think the mingw build should be restored. I wanted to raise awareness that the Build Tools include ATL/MFC, because I wasted 2 days trying to figure out things... =)

**Comment by donho** [comment]
@d0vgan Sorry for the late review. Could you solve the conflict please?
Also please use lock_guard instead please?

**Comment by donho** [comment]
@d0vgan Never mind, I will take in charge of it. Thank you for the PR.

**Pull request closed without merging**